Client/WebFilters/CosmeticFiltersResourceDownloader.swift

```diff
@@ -0,0 +1,271 @@
+    }
+    
+    /// Initialized with year 1970 to force adblock fetch at first launch.
+    private(set) var lastFetchDate = Date(timeIntervalSince1970: 0)
+    
+    func startLoading() {
+        let now = Date()
+        let fetchInterval = AppConstants.buildChannel.isPublic ? 6.hours : 10.minutes
+        
+        if now.timeIntervalSince(lastFetchDate) >= fetchInterval {
+            lastFetchDate = now
+            
+            // MUST re-create the engine otherwise we get insane load times when calling:
+            // `engine_add_resources`
+            // This is because `engine_add_resources` will ADD resources, and not delete old ones
+            // Thus we get a huge amount of memory usage and slow down.
+            engine = AdblockRustEngine()
+            
+            loadDownloadedFiles()
+            downloadCosmeticSamples()
+            downloadResourceSamples()
+        }
+    }
+    
+    func cssRules(for url: URL) -> String? {
+        engine.cssRules(for: url)
+    }
+    
+    private func loadDownloadedFiles() {
+        let fm = FileManager.default
+        guard let folderUrl = fm.getOrCreateFolder(name: CosmeticFiltersResourceDownloader.folderName) else {
+            log.error("Could not get directory with .dat and .json files")
+            return
+        }
+        
+        let enumerator = fm.enumerator(at: folderUrl, includingPropertiesForKeys: nil)
+        let filePaths = enumerator?.allObjects as? [URL]
+        let datFileUrls = filePaths?.filter { $0.pathExtension == "dat" }
+        let jsonFileUrls = filePaths?.filter { $0.pathExtension == "json" }
+        
+        datFileUrls?.forEach {
+            let fileName = $0.deletingPathExtension().lastPathComponent
+            if let data = fm.contents(atPath: $0.path) {
+                setDataFile(data: data, id: fileName)
+            }
+        }
+        
+        jsonFileUrls?.forEach {
+            let fileName = $0.deletingPathExtension().lastPathComponent
+            if let data = fm.contents(atPath: $0.path) {
+                setJSONFile(data: data, id: fileName)
+            }
+        }
+    }
+    
+    private func downloadCosmeticSamples() {
+        downloadResources(type: .cosmeticSample,
+                          queueName: "CSS Queue").uponQueue(.main) {
+            log.debug("Downloaded Cosmetic Filters CSS Samples")
+            Preferences.Debug.lastCosmeticFiltersCSSUpdate.value = Date()
+        }
+    }
+    
+    private func downloadResourceSamples() {
+        downloadResources(type: .resourceSample,
+                          queueName: "Scriplets Queue").uponQueue(.main) {
+            log.debug("Downloaded Cosmetic Filters Scriptlets Samples")
+            Preferences.Debug.lastCosmeticFiltersScripletsUpdate.value = Date()
+        }
+    }
+    
+    private func downloadResources(type: CosmeticFilterType, queueName: String) -> Deferred<()> {
+        let completion = Deferred<()>()
```

Comment:
This is fine, if you have some extra time you could consider converting it into async/await or something other than the `Deferred` lib.
I think over time we will want to deprecate this lib

Reply:
Fixed.